[WIP] Adapt recar to the unminified version of Racer

avm99963 opened the pull request:
The Racer website has changed. The scripts are now unminified, and some
pieces of code have been commented by the devs for some reason, breaking
some functionality.

This change adapts the download script so it continues working, along
with a new patch file (which is based in the old one with added
patches).

Fixes: #1

avm99963 committed:
[WIP] Adapt recar to the unminified version of Racer

The Racer website has changed. The scripts are now unminified, and some
pieces of code have been commented by the devs for some reason, breaking
some functionality.

This change adapts the download script so it continues working, along
with a new patch file (which is based in the old one with added
patches).

Fixes: #1

avm99963 committed:
Replace download.sh by download_new.sh

avm99963 commented:
Hi!

Some comments about this PR to get you on track: :)

- As of now some static resources are retrieved from https://storage.googleapis.com/chrome-racer-static/. To do: as the old download script did, these resources should be downloaded from this bucket and loaded locally, without depending on that storage bucket. IIRC, the list of resources has changed.

- Read the comment I added here: (basically, I removed a line from the code some months ago for some reason and I don't remember why, but keeping it doesn't seem to affect anything)

https://github.com/Technohacker/recar/blob/9b0efb4a34d4a51f256a84abe215dd229bb5cf37/patches/002-client-new-fix.patch#L128:L133

- Check out the updated [//README.md](https://github.com/Technohacker/recar/blob/9b0efb4a34d4a51f256a84abe215dd229bb5cf37/README.md). If you follow its instructions, you should get Racer working fine (in the JS console you'll see some errors though, maybe we'll have to sort those out?).

- Also, take a look at [//patches/experiments/README.md](https://github.com/Technohacker/recar/blob/9b0efb4a34d4a51f256a84abe215dd229bb5cf37/patches/experimentation/README.md).

Technohacker commented:
Thank you so much for the work so far! I'll have a go through the changes too :D

Technohacker commented:
I'll edit this message as I go through the changes for any overall comments :)

1. ~~The wget scraper method will miss a few resources since they won't be part of the HTML file that's scraped. Not too big a worry, we just need to add a few special cases for things like the sound effects (mainly `/assets/sounds/config.xml`. It describes where every other sound is available). This also seems to be a solution to the `__DATA__` constant being missing since those files are under `assets/data`~~ Ah my bad, you did mention it already :sweat_smile: 
2. Saving it as one giant patch would probably be unnecessary. I'd prefer the multi-file patches so it's easier to maintain for later :)

avm99963 commented:
Some replies to your comments:

1. Still, I think switching from using wget with the recursive flag to fetching the resources individually with a list like before would be great, since this would make the script more reliable. What do you think?
2. Do you mean splitting `patches/002-client-new-fix.patch` into several patch files, one per file? If this is what you meant, I agree with you!

Technohacker commented:
1. Recursive on the main page with a separate list for any JS-loaded resources will do, I've built a preliminary version on my copy and I'll push it to this branch in a while if it's alright with you :D 
2. Yup! The only reason I made my changes all in one patch was they were pretty much all related :sweat_smile: 

Some of the music tracks don't seem to load, either they aren't getting downloaded (.m4a) or they aren't loaded correctly :thinking:

avm99963 commented:
1. Awesome, go ahead! :)
2. Perfect then, I'll do that.

About the music tracks: while racing I found out some music layers were not playing, but I fixed it by adding `webaudio: true` here:

https://github.com/Technohacker/recar/blob/9b0efb4a34d4a51f256a84abe215dd229bb5cf37/mem_db.js#L20-L34

Which tracks do you find missing? I'll try to see if I'm experiencing the same too.

Technohacker committed:
Add script to download resources not listed in HTML

Technohacker commented:
Here's what I got from the JS log for the sounds. The .m4a files didn't seem to download when I was trying them, I might need to do them separately. But the ones that say the sound isn't loaded are present and accounted for https://pastebin.com/cZrG3ic7

Also, to make the client use the downloaded copy, this line will need to be changed to point to the racer folder on the server with a trailing forward slash https://github.com/Technohacker/recar/pull/3/files#diff-90b1d50f86e76f341521b5e9f6d35149ea352d5cfeaf20db3220ceb545b6aa1eR170 :)

(oh, and I tested this on desktop with `#forcemobile` at the end of the path)

avm99963 commented on download.sh:
```diff
@@ -1,181 +1,25 @@
+    cd client &&
+    # nH removes the hostname, cut-dirs removes the racer subdirectory
+    wget -r --no-parent --compression=auto --no-cookies -nH --cut-dirs=1 https://chrome.com/racer/index.html &&
+    wget -r --no-parent --compression=auto --no-cookies -nH --cut-dirs=1 --base=https://chrome.com/racer/ --input-file=../extra_resources.txt
```
Thanks for improving my code, this is much better!

Maybe `../` should be removed from `--input-file=../extra_resources.txt`?

Technohacker replied:
That one's intentional, it's in the client folder at that moment :D

avm99963 replied:
Oh wow, true 😅  Thanks!

avm99963 commented:
Just a quick note: `#forcemobile` disables sounds (see `client/assets/js/racer/config/Config.js`):

```js
  if (window.location.hash.strpos("forcemobile")) {
    this.DISABLE_SOUNDS = true;
    this.FORCE_MOBILE = true;
  }
```

Anyways, it seems like you didn't use `#forcemobile` when you recorded the console logs, because when disabling sound, it doesn't even try playing any audio.

Sound works for me as of commit 9b0efb4, so I also think it might have to do with the fact that those files haven't been downloaded locally. How did you generate the list `extra_resources.txt`?

Technohacker commented:
Oh crap, for some reason it does play most of the other sounds, I'll try fixing my mobile test setup xD

The images I copied from one of the source files, there was a neat list of them. For sounds I found the sounds/config.xml file from the official client, then for every sound file in that list I built the list as relative files to that config.xml file :)

A more sure-fire way would be using the network inspector when running it in the browser, that would probably help fill any remaining gaps :D

avm99963 commented:
Cool! I didn't remember there was a list of assets, that's quite convenient hehe.

> A more sure-fire way would be using the network inspector when running it in the browser, that would probably help fill any remaining gaps :D

Yep, I don't feel there's any other better way of doing this.

If you need any help fixing the sound please tell me!

avm99963 commented on extra_resources.txt:
```diff
@@ -0,0 +1,154 @@
+assets/js/lib/paper.js
+assets/js/lib/paper.js
+assets/data/desktop.js
+assets/data/desktop.js
```
Some additional assets to add:

```
assets/sprites/crash/blue.png
assets/sprites/crash/red.png
assets/sprites/crash/yellow.png
assets/sprites/crash/green.png
assets/sprites/crash/orange.png
assets/sprites/sparks/1.png
assets/sprites/sparks/2.png
assets/sprites/sparks/3.png
assets/sprites/sparks/4.png
assets/sprites/sparks/5.png
assets/images/content/hands.jpg
assets/images/common/logos/youtube.png
assets/videos/hands.mp4
assets/videos/hands.webm
```

Technohacker commented:
Actually, this PR as it is right now does solve the client downloading issue, so I'm willing to merge it :)
Any other blocking issue remaining?

avm99963 committed:
Remove patches/experimentation folder

avm99963 commented:
I just deleted the experimentation folder, so I guess it would be fine to merge this! :)

Your last commit adds some files which are downloaded but right now the list is not exhaustive and `Config.PATH` still points to Google Storage instead of the local server. I guess you'd like to merge it as it is and fix it in a follow-up commit? (I'm fine with this, but I just wanted to note it.)

Also, something we should do is separate the patch into individual patches as you said, but it's quite cumbersome to do. This is because for some files I've changed some things which would need to be in separate patches based on what the changes do (whether it fixes sound issues, if it changes the configuration, etc.), and doing it manually takes a considerable amount of manual work (the lines changed are very close to each other and that means having to patch an already patched file). I've realized I don't like working with patches :P

Technohacker commented:
Ah kay I'll merge this now, the rest can be moved to a new commit for sure :D 

Ah yeah I _really_ need to find a better way to manage those patches, maybe initialize a git repo there, and make git generate it for us xD

Technohacker commented:
Thank you so much for the PR again! :grin:
